File: card-api/src/main/java/com/jean202/cardmizer/api/demo/DemoScenarioController.java
package com.jean202.cardmizer.api.demo;

import java.util.List;
import java.util.Set;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api/demo-scenarios/recommendations")
public class DemoScenarioController {
    private final RecommendationDemoScenarios recommendationDemoScenarios;

    public DemoScenarioController(RecommendationDemoScenarios recommendationDemoScenarios) {
        this.recommendationDemoScenarios = recommendationDemoScenarios;
    }

    @GetMapping
    public List<RecommendationDemoScenarioResponse> list() {
        return recommendationDemoScenarios.scenarios().stream()
                .map(RecommendationDemoScenarioResponse::from)
                .toList();
    }

    public record RecommendationDemoScenarioResponse(
            String id,
            String title,
            String description,
            String expectedRecommendedCardId,
            int seedRecordCount,
            RecommendationRequestTemplate request
    ) {
        static RecommendationDemoScenarioResponse from(RecommendationDemoScenario scenario) {
            return new RecommendationDemoScenarioResponse(
                    scenario.id(),
                    scenario.title(),
                    scenario.description(),
                    scenario.expectedRecommendedCardId(),
                    scenario.seedRecords().size(),
                    RecommendationRequestTemplate.from(scenario.request())
            );
        }
    }

    public record RecommendationRequestTemplate(
            String spendingMonth,
            long amount,
            String merchantName,
            String merchantCategory,
            Set<String> paymentTags
    ) {
        static RecommendationRequestTemplate from(DemoRecommendationRequest request) {
            return new RecommendationRequestTemplate(
                    request.spendingMonth(),
                    request.amount(),
                    request.merchantName(),
                    request.merchantCategory(),
                    request.paymentTags()
            );
        }
    }
}

File: card-api/src/main/kotlin/com/jean202/cardmizer/api/demo/DemoScenarioController.kt
package com.jean202.cardmizer.api.demo

import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/demo-scenarios/recommendations")
class DemoScenarioController(
    private val recommendationDemoScenarios: RecommendationDemoScenarios,
) {
    @GetMapping
    fun list(): List<RecommendationDemoScenarioResponse> =
        recommendationDemoScenarios.scenarios.map { RecommendationDemoScenarioResponse.from(it) }

    data class RecommendationDemoScenarioResponse(
        val id: String,
        val title: String,
        val description: String,
        val expectedRecommendedCardId: String,
        val seedRecordCount: Int,
        val request: RecommendationRequestTemplate,
    ) {
        companion object {
            fun from(scenario: RecommendationDemoScenario) = RecommendationDemoScenarioResponse(
                id = scenario.id,
                title = scenario.title,
                description = scenario.description,
                expectedRecommendedCardId = scenario.expectedRecommendedCardId,
                seedRecordCount = scenario.seedRecords.size,
                request = RecommendationRequestTemplate.from(scenario.request),
            )
        }
    }

    data class RecommendationRequestTemplate(
        val spendingMonth: String,
        val amount: Long,
        val merchantName: String,
        val merchantCategory: String?,
        val paymentTags: Set<String>,
    ) {
        companion object {
            fun from(request: DemoRecommendationRequest) = RecommendationRequestTemplate(
                spendingMonth = request.spendingMonth,
                amount = request.amount,
                merchantName = request.merchantName,
                merchantCategory = request.merchantCategory,
                paymentTags = request.paymentTags,
            )
        }
    }
}
